### PR TITLE
Improve fee and change logic

### PR DESCRIPTION
**Pull request by AloeareV**
Still in progress...some tests are now failing.

### CODE REVIEW

File: zingolib/src/wallet/data.rs [hunk 1]
```diff
@@ -944,8 +945,27 @@ impl TransactionMetadata {
...
 
-    pub fn get_transaction_fee(&self) -> u64 {
-        self.total_value_spent() - (self.value_outgoing() + self.total_change_returned())
+    /// Returns Some(fee) for outgoing transactions, and None for non-outgoing
+    pub fn get_transaction_fee<P: Parameters>(
+        &self,
+        wc: &WalletCapability,
+        params: &P,
+    ) -> Option<u64> {
+        if self.is_outgoing_transaction() {
```
Comment (fluidvanadium) [inline zingolib/src/wallet/data.rs:954]: this _is_ a fix for at-least one do_list_txsummaries bug

File: zingolib/src/wallet/data.rs [hunk 2]
```diff
@@ -1145,9 +1165,45 @@ impl TransactionMetadata {
         self.value_spent_by_pool().iter().sum()
     }
 
-    pub fn value_outgoing(&self) -> u64 {
+    pub fn value_outgoing<P: Parameters>(&self, wc: &WalletCapability, params: &P) -> u64 {
+        let viewkey = wc.ufvk().unwrap();
         self.outgoing_tx_data
             .iter()
+            .filter(
```
Comment (fluidvanadium) [inline zingolib/src/wallet/data.rs:1172]: after some study, i have concluded this filters away send to self and only passes outgoing transactions. if thats true, can we add a comment explaining it?

**Review by fluidvanadium — CHANGES_REQUESTED** [review]

File: zingolib/src/wallet/data.rs (hunk 2 above)
Comment (AloeareV) [reply zingolib/src/wallet/data.rs:1174]: Done!

**Review by fluidvanadium — APPROVED** [review]

**Review by fluidvanadium — APPROVED** [review]
darn, i wish i could have beaten you to a failing test, but this should fix

**Review by fluidvanadium — CHANGES_REQUESTED** [review]
wait hold on, i discovered something with a test that doesnt add up to my understanding!

**Comment by AloeareV** [comment]
> okay, heres what in the regtest case, calling get_transaction_fee :
> 
> ```
> [zingolib/src/wallet/data.rs:948] self.is_outgoing_transaction() = false
> [zingolib/src/wallet/data.rs:949] self.total_value_spent() = 0
> [zingolib/src/wallet/data.rs:950] self.value_outgoing() = 0
> [zingolib/src/wallet/data.rs:950] self.total_change_returned() = 0
> [zingolib/src/wallet/data.rs:948] self.is_outgoing_transaction() = true
> [zingolib/src/wallet/data.rs:949] self.total_value_spent() = 625000000
> [zingolib/src/wallet/data.rs:950] self.value_outgoing() = 1000
> [zingolib/src/wallet/data.rs:950] self.total_change_returned() = 624989000
> [zingolib/src/wallet/data.rs:948] self.is_outgoing_transaction() = true
> [zingolib/src/wallet/data.rs:949] self.total_value_spent() = 625000000
> [zingolib/src/wallet/data.rs:950] self.value_outgoing() = 1000
> [zingolib/src/wallet/data.rs:950] self.total_change_returned() = 624989000
> [zingolib/src/wallet/data.rs:948] self.is_outgoing_transaction() = false
> [zingolib/src/wallet/data.rs:949] self.total_value_spent() = 0
> [zingolib/src/wallet/data.rs:950] self.value_outgoing() = 0
> [zingolib/src/wallet/data.rs:950] self.total_change_returned() = 0
> [zingolib/src/wallet/data.rs:948] self.is_outgoing_transaction() = false
> [zingolib/src/wallet/data.rs:949] self.total_value_spent() = 0
> [zingolib/src/wallet/data.rs:950] self.value_outgoing() = 0
> [zingolib/src/wallet/data.rs:950] self.total_change_returned() = 0
> [zingolib/src/wallet/data.rs:948] self.is_outgoing_transaction() = false
> [zingolib/src/wallet/data.rs:949] self.total_value_spent() = 0
> [zingolib/src/wallet/data.rs:950] self.value_outgoing() = 0
> [zingolib/src/wallet/data.rs:950] self.total_change_returned() = 0
> ```
`value_outgoing` doesn't include the transaction fee, the fee is what's left when you subtract the value outgoing + change returned from the value spent. The behavior is correct, it's the name that's misleading. Any idea on what to call this function so that it's clear it doesn't include the fee?

**Comment by fluidvanadium** [comment]
non-outgoing transactions return all zeros... but i suspect this has not been the case in real wallets. leaving an open question

what conditions cause total_value_spent() to change?

**Comment by AloeareV** [comment]
> what conditions cause TransactionMetadata.total_change_returned() to change?
> 
> ```
> TransactionMetadata.is_change()
> ```
> 
> i have not figured out where the latter is written

The last one is written by transactions.rs `mark_notes_as_change_for_pool`

**Comment by fluidvanadium** [comment]
from this branch, zingocli crashes on summaries like so:
```
TXID 26b77aa06cd6d765783281ce1cb34b2a8b9cba7c2876094ac8c913937f722b4f spent 53263788993, received 53263777993, and send out 1000
TXID 591c686cca852bc388c0eaee15e18e3f9b11d8f5adaa302942f80d0457cf1ee6 spent 96394571129, received 79194561129, and send out 17200000000
TXID f88ef9107b08817199b0f55ed3d3648841ccef7d7cc208b3fd075e51e4394086 spent 18063706992, received 3063696992, and send out 30000000000
thread '<unnamed>' panicked at 'attempt to subtract with overflow', zingolib/src/wallet/data.rs:963:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error executing command summaries: receiving on a closed channel
```

**Comment by fluidvanadium** [comment]
why does outgoing require parameters?

**Comment by AloeareV** [comment]
> why does outgoing require parameters?

The parameters specify the human-readable part of the address encoding, which we need to check if the address string we sent to is derivable from our keys

**Pull request closed without merging**